staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -668,22 +685,19 @@ func (r *Reflector) watchList(stopCh <-chan struct{}) (watch.Interface, error) {
 			}
 			return nil, err
 		}
-		bookmarkReceived := pointer.Bool(false)
-		err = watchHandler(start, w, temporaryStore, r.expectedType, r.expectedGVK, r.name, r.typeDescription,
+		watchListBookmarkReceived, err := handleListWatch(start, w, temporaryStore, r.expectedType, r.expectedGVK, r.name, r.typeDescription,
 			func(rv string) { resourceVersion = rv },
-			bookmarkReceived,
 			r.clock, make(chan error), stopCh)
 		if err != nil {
-			w.Stop() // stop and retry with clean state
```

Comment:
How can we just delete this `Stop()` without any replacement for it?

Reply:
Good catch. I had ripped it out and moved it to a defer, but the defer is part of the bug fix, so it got pulled out of this "just refactor" pr. I'll add it back in. I'm surprised so many tests passed without it, but I guess that just goes to show that the tests aren't really verifying closure of the channels.

Reply:
Adding it back in will probably fix the "EtcdMain goroutine check" integration test, but that means none of the unit tests are validating that the result channel is being closed when the stop channel is closed or on retriable error.

Reply:
Hmm. Did not in fact fix the "EtcdMain goroutine check" error. So maybe I broke something else somewhere...

Reply:
Ah. I think the extra stop checks I added don't work well without the added defers. So it's actually making the problem worse and not closing the result channel in even more cases now. So I'll probably need to rip those out and only add them with the defers in the next PR.

Reply:
Instead of removing the early stop channel checks, I added w.Stop() calls to them. Hopefully that fixes the etcd goroutine leak.